:art: Refactor `ModelABC` to Help Use Default Torch Models

shaneahmed opened the pull request:
- Refactor `ModelABC` to Help Use Default Torch Models
- Remove `infer_batch` from `ModelABC`

shaneahmed committed:
:wrench: Use `pyproject.toml` for `bdist_wheel` configuration
- Use `pyproject.toml` for `bdist_wheel` configuration

shaneahmed committed:
Merge remote-tracking branch 'origin/develop' into dev-define-engines-abc

shaneahmed committed:
:zap: Improve `Engines` performance and implementation
- Improve `Engines` performance and implementation

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:recycle:  Refactor engines_abc.py
- Refactor engines_abc.py

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:recycle: Refactor base code from `IOSegmentorConfig` to `ModelIOConfigABC` (#618)

- Moved all ioconfigs to a single file.
- Used dataclass to define ioconfig.
- Refactor base code from `IOSegmentorConfig` to `ModelIOConfigABC`
- Use `ModelIOConfigABC` for `PatchPredictor` instead of `IOSegmentorConfig`

---------

Signed-off-by: Shan E Ahmed Raza <13048456+shaneahmed@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: John Pocock <John-P@users.noreply.github.com>
Co-authored-by: Mark Eastwood <20169086+measty@users.noreply.github.com>
Co-authored-by: Mostafa Jahanifar <74412979+mostafajahanifar@users.noreply.github.com>
Co-authored-by: Adam Shephard <39619155+adamshephard@users.noreply.github.com>

shaneahmed committed:
:recycle: Move Dataset Classes to `dataset_abc.py` (#637)

- Move `WSIStreamDataset`, `PatchDatasetABC` and `WSIPatchDataset` to `dataset_abc.py`

---------

Signed-off-by: Shan E Ahmed Raza <13048456+shaneahmed@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:twisted_rightwards_arrows: Merge develop into dev-define-engines-abc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
:recycle: Refactor Minor Changes to Keep #635 Clean (#705)

* :recycle: Refactor minor changes to keep #635 clean

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:twisted_rightwards_arrows: Merge develop into dev-define-engines-abc

shaneahmed committed:
:twisted_rightwards_arrows: Merge develop into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:recycle: Move `DataSet` tests to `test_dataset.py` (#708)

-  Move `DataSet` tests to `test_dataset.py`

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:bug: Fix `tiatoolbox/models/dataset/classification.py` for annotations

shaneahmed committed:
:bug: Fix `tiatoolbox/models/dataset/classification.py` for annotations

shaneahmed committed:
:bug: Fix `tiatoolbox/models/dataset/dataset_abc.py` for annotations

shaneahmed committed:
:twisted_rightwards_arrows: Merge `develop` into `dev-define-engine-abc`

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:art: New `EngineABC` Design for `Patches` (#635)

- New `EngineABC` design
- Update `PatchPredictor` based on the new design
- Move `model_to` from `utils.misc` to `model_abc`
- Define `load_torch_model` in `model_abc`

---------

Signed-off-by: Shan E Ahmed Raza <13048456+shaneahmed@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: measty <20169086+measty@users.noreply.github.com>
Co-authored-by: abishekrajvg <abishekraj6797@gmail.com>

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:twisted_rightwards_arrows: Merge develop into dev-define-engines-abc

Signed-off-by: Shan E Ahmed Raza <13048456+shaneahmed@users.noreply.github.com>

shaneahmed committed:
:goal_net: Fix merge errors in classification.py

Signed-off-by: Shan E Ahmed Raza <13048456+shaneahmed@users.noreply.github.com>

shaneahmed committed:
:goal_net: Fix merge errors in dataset_abc.py

shaneahmed committed:
:goal_net: Fix merge errors in dataset_abc.py

shaneahmed committed:
:goal_net: Fix test_models_abc.py

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:twisted_rightwards_arrows: Merge `develop` into `dev-define-engine-abc`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

AbishekRajVG committed:
✨ Add WSI processing to `EngineABC` (#737)

Implements cache mode for processing using EngineABC. Cache_mode saves intermediate results to zarr file which can be converted to AnnotationStore. Full WSI pipeline needs to be implemented for Engines inheriting the base class.

The output of models should be a dictionary according to the new Engine design.

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Shan E Ahmed Raza <13048456+shaneahmed@users.noreply.github.com>

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:rewind: Restore `test_patch_predictor.py`

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
🧑‍💻 Define `PatchPredictor` (#783)

- Redesigns PatchPredictor engine using the new EngineABC base class.
- The WSIs are now processed using the same code as for the processing the patches using WSI based dataloader.
- The intermediate output is saved as zarr for the WSIs to resolve memory issues.
- The output of model architectures should now be a dictionary.
- The output can be specified as AnnotationStore for visualisation using TIAViz.

---------

Co-authored-by: abishekrajvg <abishekraj6797@gmail.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:bug: Fix `mypy` Type Checks for `cli/common.py` (#864)

- Fix `mypy` Type Checks for `cli/common.py`

shaneahmed committed:
:art: Refactor `ModelABC` to Help Use Default Torch Models

Signed-off-by: Shan E Ahmed Raza <13048456+shaneahmed@users.noreply.github.com>

shaneahmed committed:
:white_check_mark: Fix test

Signed-off-by: Shan E Ahmed Raza <13048456+shaneahmed@users.noreply.github.com>

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

GeorgeBatch commented on tiatoolbox/models/architecture/vanilla.py:
```diff
@@ -78,6 +78,40 @@ def _get_architecture(
+    with torch.inference_mode():
+        output = model(img_patches_device)
+    # Output should be a single tensor or scalar
+    return {"probabilities": output.cpu().numpy()}
```
In the current `develop` branch, neither `CNNModel`, nor `CNNBackbone` returned dictionaries as output of their `infer_batch()` methods. Also, `CNNModel` currently returns an array, while `CNNBackbone` returns a list with the array. It might be fine, just wanted to highlight this.

`CNNModel`
https://github.com/TissueImageAnalytics/tiatoolbox/blob/76f02f8561313dae786ebc15ab3196bde836a441/tiatoolbox/models/architecture/vanilla.py#L173

`CNNBackbone`
https://github.com/TissueImageAnalytics/tiatoolbox/blob/76f02f8561313dae786ebc15ab3196bde836a441/tiatoolbox/models/architecture/vanilla.py#L270

shaneahmed replied:
Thanks. We are aware of this. Our preference is to use torch nn models but to generalise for multi modal output we may need dictionaries. This PR is to check if we can move to generic torch models or we will need a sub class.

GeorgeBatch replied:
Makes sense.

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
:twisted_rightwards_arrows: Merge `develop` into `dev-define-engine-abc`

shaneahmed committed:
:bug: Fix `model_to` import

shaneahmed committed:
:twisted_rightwards_arrows: Merge `develop` into `dec-define-engine-abc`

shaneahmed committed:
:bug: Fix `model_to` device specification

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

# Conflicts:
#	tests/models/test_arch_vanilla.py
#	tiatoolbox/models/architecture/vanilla.py

shaneahmed committed:
✨ Add `PatchPredictor` Engine (#865)

-  Add `PatchPredictor` Engine based on `EngineABC`
- Add `return_probabilities` option to Params
- Removes `merge_predictions` option in `PatchPredictor` engine.
- Defines `post_process_cache_mode` which allows running the algorithm on `WSI`
- Add `infer_wsi` for WSI inference
- Removes `save_wsi_output` as this is not required after post processing.
- Removes `merge_predictions` and fixes docstring in EngineABCRunParams
- `compile_model` is now moved to EngineABC init
- Fixes bug with `_calculate_scale_factor`
- Fixes a bug in `class_dict` definition.
- `_get_zarr_array` is now a public function `get_zarr_array` in `misc`
- `patch_predictions_as_annotations` runs the loop on `patch_coords` instead of `class_probs`

---------

Co-authored-by: Mark Eastwood <20169086+measty@users.noreply.github.com>
Co-authored-by: Mostafa Jahanifar <74412979+mostafajahanifar@users.noreply.github.com>
Co-authored-by: Adam Shephard <39619155+adamshephard@users.noreply.github.com>
Co-authored-by: Jiaqi-Lv <60471431+Jiaqi-Lv@users.noreply.github.com>

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

# Conflicts:
#	tiatoolbox/models/engine/engine_abc.py

shaneahmed committed:
:twisted_rightwards_arrows: Merge `develop` into `dev-define-engines-abc`

shaneahmed committed:
:twisted_rightwards_arrows: Undo unwanted changes during merge.

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
:bug: Fix PLC0206 (#907)

* :bug: Fix PLC0206
- Fix PLC0206 Extracting value from dictionary without calling `.items()`

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

# Conflicts:
#	tests/models/test_multi_task_segmentor.py
#	tiatoolbox/models/dataset/classification.py
#	tiatoolbox/models/engine/patch_predictor.py
#	tiatoolbox/models/engine/semantic_segmentor.py

shaneahmed committed:
:bug: Fix tiatoolbox type_hints import

shaneahmed committed:
:bug: Fix deepsource bugs

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
:recycle: Use `input_resolutions` instead of `resolution` in Engine Params (#917)

* :recycle: Use `input_resolutions` instead of resolution
- Use `input_resolutions` instead of resolution to make engines outputs compatible with ioconfig.
- Uses input resolution as a list of dictionaries on units and resolution.

* :recycle: Use `input_resolutions` instead of resolution
- Use `input_resolutions` instead of resolution to make engines outputs compatible with ioconfig.
- Uses input resolution as a list of dictionaries on units and resolution.

* :white_check_mark: Add test to cli.

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
:twisted_rightwards_arrows: Merge develop into dev-define-engines-abc

shaneahmed committed:
:fire: Remove unwanted import

shaneahmed committed:
:bug: Fix ModelIOConfigABC import

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

# Conflicts:
#	tiatoolbox/models/engine/patch_predictor.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

# Conflicts:
#	tiatoolbox/utils/misc.py

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

# Conflicts:
#	tests/models/test_feature_extractor.py
#	tiatoolbox/models/models_abc.py

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
:twisted_rightwards_arrows: Merge branch 'develop' into dev-define-engines-abc

# Conflicts:
#	tiatoolbox/cli/common.py
#	tiatoolbox/cli/nucleus_instance_segment.py
#	tiatoolbox/cli/patch_predictor.py
#	tiatoolbox/models/engine/semantic_segmentor.py

shaneahmed committed:
:bug: Fix FBT001

shaneahmed committed:
:bug: Fix mypy checks

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
🔀 Merge branch `develop` into `dev-engine-abc` (#948)

* ⚡ Make WSIPatchDataset Pickleable to Support Windows Multithreading (#947)

This PR makes the WSIPatchDataset class picklable by delaying the creation of the reader object until the first call to `__getitem__`. This enables the use of multiple loader workers on Windows without errors and provides significant performance improvements.

- Delays reader object instantiation to the first `__getitem__` call instead of during initialization
- Extracts reader creation logic into a separate `_get_reader` method
- Stores image path and mode as instance variables for lazy initialization

Speedup for the WSI prediction cell of the patch_prediction example notebook: 
2min 48 sec with 0 loader workers -> 1min 13 sec with 4 workers.

Note: this PR doesn't have any effect for Linux as the multi-threading already works fine there because Linux multithreading doesn't require things to be pickleable

* :twisted_rightwards_arrows: Merge branch develop into dev-engine-abc

* :bug: Fix reader_info read

---------

Co-authored-by: Mark Eastwood <20169086+measty@users.noreply.github.com>

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

# Conflicts:
#	tiatoolbox/models/dataset/classification.py

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
Merge branch 'develop' into dev-define-engines-abc

# Conflicts:
#	tests/models/test_patch_predictor.py

shaneahmed committed:
:bug: Fix Use a raw string or `re.escape()` to make the intention explicit

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model

shaneahmed committed:
:twisted_rightwards_arrows: Merge `develop` into `dev-engine-abc`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
:bug: Fix `ruff` error

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'dev-define-engines-abc' into dev-convert-modelabc-to-torch-nn-model
